presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java

```diff
@@ -187,59 +187,76 @@ private static double firstNonNaN(double... values)
-
-            StatisticRange rightRange = StatisticRange.from(rightColumnStats);
-            StatisticRange antiRange = StatisticRange.from(leftColumnStats)
-                    .subtract(rightRange);
-
-            // TODO: use NDVs from left and right StatisticRange when they are fixed
-            double leftNDV = leftColumnStats.getDistinctValuesCount();
-            double rightNDV = rightColumnStats.getDistinctValuesCount();
-
-            if (leftNDV > rightNDV) {
-                double selectedRangeFraction = leftColumnStats.getValuesFraction() * (leftNDV - rightNDV) / leftNDV;
-                double scaleFactor = selectedRangeFraction + leftColumnStats.getNullsFraction();
-                double newLeftNullsFraction = leftColumnStats.getNullsFraction() / scaleFactor;
-                outputStats = outputStats.mapSymbolColumnStatistics(clause.getLeft(), columnStats ->
-                        SymbolStatsEstimate.buildFrom(columnStats)
-                                .setLowValue(antiRange.getLow())
-                                .setHighValue(antiRange.getHigh())
-                                .setNullsFraction(newLeftNullsFraction)
-                                .setDistinctValuesCount(leftNDV - rightNDV)
-                                .build());
-                outputStats = outputStats.mapOutputRowCount(rowCount -> rowCount * scaleFactor);
+        if (criteria.isEmpty()) {
+            if (rightStats.getOutputRowCount() > 0) {
+                // all left side rows are matched
+                return new AntiJoinStats(leftStats.mapOutputRowCount(rowCount -> 0.0), ImmutableSet.of());
             }
-            else if (leftNDV <= rightNDV) {
-                // only null values are left
-                outputStats = outputStats.mapSymbolColumnStatistics(clause.getLeft(), columnStats ->
-                        SymbolStatsEstimate.buildFrom(columnStats)
-                                .setLowValue(NaN)
-                                .setHighValue(NaN)
-                                .setNullsFraction(1.0)
-                                .setDistinctValuesCount(0.0)
-                                .build());
-                outputStats = outputStats.mapOutputRowCount(rowCount -> rowCount * leftColumnStats.getNullsFraction());
+            else if (rightStats.getOutputRowCount() == 0) {
+                // none left side rows are matched
+                return new AntiJoinStats(leftStats, ImmutableSet.of());
             }
             else {
-                // either leftNDV or rightNDV is NaN
-                return UNKNOWN_STATS;
+                // right stats row count is NaN
+                return new AntiJoinStats(UNKNOWN_STATS, ImmutableSet.of());
             }
         }
 
-        return outputStats;
+        // heuristics: select the most selective criteria for anti join clause
+        return criteria.stream()
+                .map(clause -> {
```

Comment:
this function is kind of long for an anonymous function.  Can you make it not anoymous so it's easier to see the flow?